File: project/frontend/src/lib/nodes/definitions/pytorch/flatten.ts
/**
 * PyTorch Flatten Layer Node Definition
 * Enhanced with pattern-based validation
 */

import { NodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField, ShapePattern, DimensionValue } from '../../../types'
import { minRank } from '../../../validation/patterns'
import { getRank, isNumeric } from '../../../validation/matchers'

export class FlattenNode extends NodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'flatten',
    label: 'Flatten',
    category: 'basic',
    color: 'var(--color-primary)',
    icon: 'ArrowsDownUp',
    description: 'Flatten tensor dimensions',
    framework: BackendFramework.PyTorch
  }

  /**
   * Input pattern: at least 2D tensor
   */
  readonly inputPattern: ShapePattern = minRank(2)

  readonly configSchema: ConfigField[] = [
    {
      name: 'start_dim',
      label: 'Start Dimension',
      type: 'number',
      default: 1,
      min: 0,
      description: 'First dimension to flatten'
    },
    {
      name: 'end_dim',
      label: 'End Dimension',
      type: 'number',
      default: -1,
      description: 'Last dimension to flatten (-1 for last)'
    }
  ]

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    if (!inputShape) {
      return undefined
    }

    const rank = getRank(inputShape)
    const startDim = (config.start_dim as number) ?? 1
    let endDim = (config.end_dim as number) ?? -1

    // Normalize negative index
    if (endDim < 0) {
      endDim = rank + endDim
    }

    // Validate dimensions
    if (startDim < 0 || startDim >= rank || endDim < startDim || endDim >= rank) {
      return inputShape
    }

    // Get dimensions to flatten
    const dimsToFlatten = inputShape.dims.slice(startDim, endDim + 1)

    // Calculate flattened size
    let flattenedSize: DimensionValue
    if (dimsToFlatten.every(isNumeric)) {
      flattenedSize = (dimsToFlatten as number[]).reduce((a, b) => a * b, 1)
    } else {
      // Symbolic - create expression
      flattenedSize = dimsToFlatten.map(d => String(d)).join('*')
    }

    // Build output shape
    const outputDims: DimensionValue[] = [
      ...inputShape.dims.slice(0, startDim),
      flattenedSize,
      ...inputShape.dims.slice(endDim + 1)
    ]

    return {
      dims: outputDims,
      description: `Flattened dims ${startDim} to ${endDim}`,
      flags: {
        inferred: true,
        symbolic: !isNumeric(flattenedSize)
      },
      provenance: {
        source: 'computed',
        transformation: 'flatten',
        description: `${inputShape.dims.join('×')} → ${outputDims.join('×')}`
      }
    }
  }

  getDefaultConfig(): BlockConfig {
    return {
      start_dim: 1,
      end_dim: -1
    }
  }
}

File: project/block_manager/services/nodes/frameworks/pytorch/nodes/flatten.py
"""PyTorch Flatten Layer Node Definition"""

from typing import Any, Dict, List, Optional

from ....core.types import ConfigField, Framework, NodeMetadata, TensorShape
from ..base import PyTorchNodeDefinition


class FlattenNode(PyTorchNodeDefinition):
    """Flatten multi-dimensional input to 2D"""

    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="flatten",
            label="Flatten",
            category="basic",
            color="var(--color-primary)",
            icon="ListBullets",
            description="Flatten multi-dimensional input to 2D",
            framework=Framework.PYTORCH
        )

    @property
    def config_schema(self) -> List[ConfigField]:
        return []  # No configuration needed

    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        if not input_shape:
            return None

        if len(input_shape.dims) < 2:
            return None

        # Calculate flattened size (multiply all dims except batch)
        batch = input_shape.dims[0]
        flat_size = 1
        for dim in input_shape.dims[1:]:
            flat_size *= dim

        return TensorShape(
            dims=[batch, flat_size],
            description="Flattened features"
        )

    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        return None  # Accept any input
